app/main/presenters/search_presenters.py

```diff
@@ -91,45 +92,99 @@ def set_filter_states(filter_groups, request):
                 )
 
 
-def annotate_lots_with_categories_selection(lots, category_filter_group, request):
+def _get_category_filter_key_set(category_filter_group):
+    """
+    Returns the set of keys used in the category filter group. In practice
+    there should only be one, but for completeness we allow for the possibility
+    of there being several. In G6/7/8, this was {'serviceTypes'}, in G9 it's
+    {'serviceCategories'} .
+    :return: set[string]
+    """
+    if category_filter_group is not None and category_filter_group['filters']:
+        return {f['name'] for f in category_filter_group['filters']}
+    else:
+        return set()
+
+
+def build_lots_and_categories_link_tree(lots, category_filter_group, request):
     """
     Equivalent of set_filter_states but for where we are creating a tree of links i.e. the
     lots/categories widget. Adds links (where necessary) and shows the currently-selected
     lot and/or categories.
-    :param lots: a sequence of lot dicts, which this function will annotate
+
+    Returns a list of selected filters: there will always be at least one element, which is
+    the root of the tree; there may then be a category; there may then follow a sub-category.
+
+    :param lots: a sequence of lot dicts applicable to the live framework(s)
     :param category_filter_group: a single filter group loaded from the framework search_filters manifest
     :param request: current request so that we can figure out what lots and categories are selected
-    :return: the category filter that was directly selected, if any
+    :return: list of selected category and lot filters, starting with the 'all categories' root node node
     """
     current_lot_slug = get_lot_from_request(request)
-    selected_category = None
+
+    # Links in the tree should preserve all the filters, except those relating to this tree (i.e. lot
+    # and category).
+    search_link_builder = Href(url_for('.search_services'))
+    keys_to_remove = _get_category_filter_key_set(category_filter_group)
+    keys_to_remove.add('lot')
+    keys_to_remove.add('page')
```

Comment:
I wonder if we can do this non-magically, e.g. by removing/preserving all args that aren't used as filters - just in case we have or add something else in the future that shouldn't be affected.

Lot would probably still have to be magical for the time being.